$^{(1)}$ & External sample name\\ 
 $^{(2)}$ & Number of bases in the reference genome used for alignment\\ 
 $^{(3)}$ & Number of bases which have one or more baits on top of them\\ 
 $^{(4)}$ & Unique number of target bases in the experiment where target is usually exons etc\\ 
 $^{(5)}$ & Total number of reads including all PF and non-PF reads\\ 
 $^{(6)}$ & Number of bases in the PF aligned reads that are aligned to the reference genome. Accounts for clipping and gaps (=indels?)\\ 
 $^{(7)}$ & Mean coverage of all baits\\ 
 $^{(8)}$ & Mean coverage of targets with coverage $\ge$ 2\\ 
 $^{(9)}$ & Fraction bp near bait (+/- 250 bp)\\ 
 $^{(10)}$ & Capture specificity; 1 - fraction bases that is off bait\\ 
 $^{(11)}$ & Percentage of ALL target bases with coverage $\ge$ 2\\ 
 $^{(12)}$ & Percentage of ALL target bases with coverage $\ge$ 10\\ 
 $^{(13)}$ & Percentage of ALL target bases with coverage $\ge$ 20\\ 
 $^{(14)}$ & Percentage of ALL target bases with coverage $\ge$ 30\\ 
 $^{(15)}$ & Fraction of aligned, deduplicated, on-target bases out of the PF bases available\\ 
 $^{(16)}$ & Mean read length of the set of reads examined\\ 
 $^{(17)}$ & Number of PF reads aligned to the positive strand of the genome divided by the number of PF reads aligned to the genome\\ 
 $^{(18)}$ & Median of the insert sizes, where insert refers to the base pairs between the adapters.\\ 
 $^{(19)}$ & Mean of the insert sizes, where insert refers to the base pairs between the adapters.\\ 
 $^{(20)}$ & Standard deviation of the insert sizes, where insert refers to the base pairs between the adapters.\\ 
 $^{(21)}$ & Fraction of SNPs that are concordant between array and NGS\\ 
 
